### DISCUSSION TITLE
Mathematical functions : log and ln

### DISCUSSION THREAD

**Discussion by navy-cap-echo** [godotengine/godot-proposals#5990] [General Discussions]
Hello

Again today, I spent far too much time debugging a "mathematical" godot program that seemed incorrect but was not due to the use of the "**_log_**" function !

As specified in the documentation, the _**log**_ function :

> "Returns the natural logarithm of x. This is the amount of time needed to reach a certain level of continuous growth.
    **Note**: This is not the same as the "log" function on most calculators, which uses a base 10 logarithm."

As a reminder : 

`log(x) = ln(x)/ln(10)`

So, i should replace all my **_log_** calls by `log(x)/log(10)`, which will make the code even more confusing !

To add clarity, couldn't we just rename this function to _**ln**_ (natural logarithm), which it really is ? and add the classic _**log**_ function ?

**Comment by AThousandShips** [comment]
This is the convention in the standard library in c which is what Godot uses, but adding log10 would make sense, python math library also uses log to mean natural logarithm, so changing log to mean base 10 would be confusing and compatibility breaking, confusing as the calculator convention isn't the only one and many programming languages don't use the same conversation

**Reply by navy-cap-echo** [reply]
Aren't conventions made to be challenged ? :P

A `log10` function could do the trick, but what about adding an optional parameter for the base (By default, the base is _e_) ? 
`log(X, base)`

Then there would be no break in compatibility, and more flexibility.

**Reply by theraot** [reply]
There are three commonly used logarithms: base e, base 10 and base 2. They could use dedicated functions: `ln`, `log10`, `log2`. And then have `log` with a parameter for any other base. *However I don't expect this to get much traction when you can define them in your code with division of logarithms.*

**Reply by navy-cap-echo** [reply]
> However I don't expect this to get much traction when you can define them in your code with division of logarithms.

@theraot With this reasoning, maybe we should also remove the function `tan` which is a division of `sin` and `cos` ! 
... Just kidding.

There has always been a great deal of confusion with the log function, as specified by Wolfram :
[Wolfram Mathworld natural logarithm](https://mathworld.wolfram.com/Ln.html)
> The United States Department of Commerce recommends that the notation ln<sub>x</sub> be used in this way to refer to the natural logarithm (Taylor 1995, p. 33).
**_Unfortunately_**, mathematicians in the United States commonly use the symbol log x to refer to the natural logarithm, as does TraditionalForm typesetting in the Wolfram Language. The use of log<sub>x</sub> for different purposes by different mathematical communities _**causes considerable confusion**_, so extreme care is needed in determining if the symbol  log x found in the wild refers to log<sub>2</sub>x or log<sub>10</sub>x.

IMHO, assigning function names to be less error prone, when using them, should be a priority.  
By adding the function `ln` and adding an optional parameter for the base to the function `log`, you are removing a lot of the confusion.
